### PR TITLE
drivers: w1: add zephyr-gpio driver

### LINKED ISSUE #65787: Add GPIO as the link layer for the 1-Wire bus
**Is your feature request related to a problem? Please describe.**
At the moment, the implementation of the 1-Wire bus requires using the UART as the link layer.

**Describe the solution you'd like**
The possibility to choose to a different link layer, i.e. GPIO.

**Describe alternatives you've considered**
I've considered using a I2C multiplexer to connect the UART to all my sensors, but that requires additional cost and changes in the hardware. I've also considered implementing the 1-Wire through GPIO without integrating into Zephyr, but then I would not be able to use the existing DS18B20 library.

**Additional context**
I am developing a solution using nRF5340 and DS18B20 sensors, where I need to know the physical location of each sensor. To achieve this, I have connected each sensor to a different GPIO to map the location to the pin. But that means I can not use the provided DS18B20 library because the underlying link layer requires using the UART peripheral.

I am willing to develop the GPIO link layer and create a PR to merge with Zephyr. But since it is my first contribution to the project, I will definitely need guidance for this.

My initial thought was to implement the following functions found at `zephyr/include/zephyr/drivers/w1.h`

``` c
typedef int (*w1_reset_bus_t)(const struct device *dev);
typedef int (*w1_read_bit_t)(const struct device *dev);
typedef int (*w1_write_bit_t)(const struct device *dev, bool bit);
typedef int (*w1_read_byte_t)(const struct device *dev);
typedef int (*w1_write_byte_t)(const struct device *dev, const uint8_t byte);
typedef int (*w1_read_block_t)(const struct device *dev, uint8_t *buffer,
			       size_t len);
typedef int (*w1_write_block_t)(const struct device *dev, const uint8_t *buffer,
				size_t len);
typedef size_t (*w1_get_slave_count_t)(const struct device *dev);
typedef int (*w1_configure_t)(const struct device *dev,
			      enum w1_settings_type type, uint32_t value);
typedef int (*w1_change_bus_lock_t)(const struct device *dev, bool lock);
```

Something similar to what have been done at `drivers/w1/w1_zephyr_serial.c`, but maybe create a new file `drivers/w1/w1_zephyr_gpio.c` ?

### PR DESCRIPTION
**Pull request by hcd-bdltd**
Added GPIO as the link layer for the 1-Wire bus (#65787).

### CODE REVIEW

File: drivers/w1/Kconfig.zephyr_gpio
Comment (pdgendt) [inline drivers/w1/Kconfig.zephyr_gpio:10]: No need for `DT_COMPAT_ZEPHYR_GPIO`
```suggestion
	default y
	depends on DT_HAS_ZEPHYR_W1_GPIO_ENABLED
```

File: drivers/w1/w1_zephyr_gpio.c
Comment (nixward) [inline drivers/w1/w1_zephyr_gpio.c:285]: For simplicity:
```suggestion
		data->timing = (value != 0) ? &od : &std;
```

**Review by nixward — COMMENTED** [review]
Nice code, it's a pity k_busy_wait_ns() isn't available for your driver.

File: drivers/w1/Kconfig.zephyr_gpio
Comment (pdgendt) [inline drivers/w1/Kconfig.zephyr_gpio:6]: Can be removed

File: drivers/w1/Kconfig.zephyr_gpio
Comment (pdgendt) [inline drivers/w1/Kconfig.zephyr_gpio:17]:
```suggestion
	  The bus reset, and bit read and write operations are executed
	  via byte read and write operations on top of the Zephyr
	  GPIO driver interface.
```

File: dts/bindings/w1/zephyr,w1-gpio.yaml
Comment (pdgendt) [inline dts/bindings/w1/zephyr,w1-gpio.yaml:6]: Why not let the driver set these?

**Review by pdgendt — COMMENTED** [review]
This driver isn't built currently in CI, it would be nice to have a `w1` test added, for example in:
`zephyr/tests/drivers/build_all/w1`

File: drivers/w1/w1_zephyr_gpio.c
Comment (pdgendt) [inline drivers/w1/w1_zephyr_gpio.c:292]: Nit: indentation

**Comment by hcd-bdltd** [comment]
> This driver isn't built currently in CI, it would be nice to have a `w1` test added, for example in: `zephyr/tests/drivers/build_all/w1`

Great, I'll do it. Can you provide some documentation about running the tests locally? So I can ensure they pass before submitting.

**Comment by pdgendt** [comment]
> Great, I'll do it. Can you provide some documentation about running the tests locally? So I can ensure they pass before submitting.

For the `build_all` folder you should only be able to build without testing. You can look at the other peripherals to see how you can add a device an `app.overlay`.
For example `test_gpio_leds` in `tests/drivers/build_all/led/app.overlay` is pretty similar.

**Comment by hcd-bdltd** [comment]
Done. I've added the `tests/drivers/build_all/w1` and validated by running `west build -p -b qemu_x86 tests/drivers/build_all/w1 -T drivers.w1.build` and `west twister -p qemu_x86 -T tests/drivers/build_all/w1`

File: drivers/w1/w1_zephyr_gpio.c
Comment (nixward) [inline drivers/w1/w1_zephyr_gpio.c:133]: This may save some RAM:

```suggestion
static const struct w1_gpio_timing std = {
	.a = W1_GPIO_TIMING_STD_A,
	.b = W1_GPIO_TIMING_STD_B,
	.c = W1_GPIO_TIMING_STD_C,
	.d = W1_GPIO_TIMING_STD_D,
	.e = W1_GPIO_TIMING_STD_E,
	.f = W1_GPIO_TIMING_STD_F,
	.g = W1_GPIO_TIMING_STD_G,
	.h = W1_GPIO_TIMING_STD_H,
	.i = W1_GPIO_TIMING_STD_I,
	.j = W1_GPIO_TIMING_STD_J,
};

static const struct w1_gpio_timing od = {
	.a = W1_GPIO_TIMING_OD_A,
	.b = W1_GPIO_TIMING_OD_B,
	.c = W1_GPIO_TIMING_OD_C,
	.d = W1_GPIO_TIMING_OD_D,
	.e = W1_GPIO_TIMING_OD_E,
	.f = W1_GPIO_TIMING_OD_F,
	.g = W1_GPIO_TIMING_OD_G,
	.h = W1_GPIO_TIMING_OD_H,
	.i = W1_GPIO_TIMING_OD_I,
	.j = W1_GPIO_TIMING_OD_J,
};
```

File: tests/drivers/build_all/w1/prj.conf
Comment (pdgendt) [inline tests/drivers/build_all/w1/prj.conf:9]: These lines aren't necessary, I think

Comment (hcd-bdltd) [reply tests/drivers/build_all/w1/prj.conf]: Correct. I've removed them.

**Review by pdgendt — DISMISSED** [review]

**Review by nixward — APPROVED** [review]

**Comment by hcd-bdltd** [comment]
Hello @str4t0m do you want to have a look and let me know what you think?

**Comment by dalprahcd** [comment]
Hello @pdgendt , @nixward and @str4t0m. What are the next steps for this PR to get merged?

**Comment by pdgendt** [comment]
> Hello @pdgendt , @nixward and @str4t0m. What are the next steps for this PR to get merged?

Getting another approval, I've pinged reviewers in the `pr-help` channel on discord to have a look.

**Review by cfriedt — DISMISSED** [review]
LGTM - I'm not a w1 maintainer, but no obvious problems.

File: drivers/w1/Kconfig.zephyr_gpio
Comment (kartben) [inline drivers/w1/Kconfig.zephyr_gpio:1]: There's a minor copy-paste error here

```suggestion
# Configuration options for the Zephyr GPIO 1-Wire Master driver
```

Comment (hcd-bdltd) [reply drivers/w1/Kconfig.zephyr_gpio]: Done.

**Review by pdgendt — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
